qabelbox/src/main/java/de/qabel/qabelbox/ui/HeaderDecoration.kt

```diff
@@ -0,0 +1,198 @@
+import android.support.v7.widget.LinearLayoutManager
+import android.support.v7.widget.RecyclerView
+import android.support.v7.widget.StaggeredGridLayoutManager
+import android.view.LayoutInflater
+import android.view.View
+
+class HeaderDecoration(private val mView: View,
+                       private val mHorizontal: Boolean,
+                       private val mParallax: Float,
+                       private val mShadowSize: Float,
+                       private val mColumns: Int) : RecyclerView.ItemDecoration() {
+
+    private val mShadowPaint: Paint?
+
+    init {
+
+        if (mShadowSize > 0) {
+            mShadowPaint = Paint()
+            mShadowPaint.shader = if (mHorizontal)
+                LinearGradient(mShadowSize, 0f, 0f, 0f,
+                        intArrayOf(Color.argb(55, 0, 0, 0), Color.argb(55, 0, 0, 0), Color.argb(3, 0, 0, 0)),
+                        floatArrayOf(0f, .5f, 1f),
+                        Shader.TileMode.CLAMP)
+            else
+                LinearGradient(0f, mShadowSize, 0f, 0f,
+                        intArrayOf(Color.argb(55, 0, 0, 0), Color.argb(55, 0, 0, 0), Color.argb(3, 0, 0, 0)),
+                        floatArrayOf(0f, .5f, 1f),
+                        Shader.TileMode.CLAMP)
+        } else {
+            mShadowPaint = null
+        }
+    }
+
+    companion object {
+
+        fun with(context: Context): Builder {
+            return Builder(context)
+        }
+
+        fun with(recyclerView: RecyclerView): Builder {
+            val layoutManager = recyclerView.layoutManager
+            if (layoutManager is GridLayoutManager) {
+                return Builder(recyclerView.context, layoutManager.spanCount)
+            } else if (layoutManager is LinearLayoutManager) {
+                return Builder(recyclerView.context, layoutManager.orientation == LinearLayoutManager.HORIZONTAL)
+            } else if (layoutManager is StaggeredGridLayoutManager) {
+                return Builder(recyclerView.context, layoutManager.spanCount)
+            } else {
+                return Builder(recyclerView.context)
+            }
+        }
+    }
+
+    override fun onDraw(c: Canvas, parent: RecyclerView, state: RecyclerView.State?) {
+        super.onDraw(c, parent, state)
+        // layout basically just gets drawn on the reserved space on top of the first view
+        mView.layout(parent.left, 0, parent.right, mView.measuredHeight)
+
+        for (i in 0..parent.childCount - 1) {
+            val view = parent.getChildAt(i)
+            if (parent.getChildAdapterPosition(view) == 0) {
+                c.save()
+                if (mHorizontal) {
+                    c.clipRect(parent.left, parent.top, view.left, parent.bottom)
+                    val width = mView.measuredWidth
+                    val left = (view.left - width) * mParallax
+                    c.translate(left, 0f)
+                    mView.draw(c)
+                    if (mShadowSize > 0) {
```

Comment:
mShadowPaint is only null when mShadowSize == 0. Combine those values in a data class, so you don't have to use !! for mShadowPaint.